### ISSUE TITLE
[PAC] Implicit casts results in no resigns in some cases

### ISSUE THREAD

**Issue by asl** [llvm/llvm-project#106487]
Consider the following code:
```c++
struct C {
  void fn() noexcept;
};

static void call(C& c, void (C::*f)()) {
  (c.*f)();
}

static void callNE(C& c, void (C::*f)() noexcept) {
  (c.*f)();
}

void ref() {
  C c;
  call(c, &C::fn); // <= implicit cast removes noexcept
  callNE(c, &C::fn);
}
```

It produced the LLVM IR as follows:
```llvm
 call void ptrauth (ptr @_ZN1C2fnEv, i32 0, i64 62113)(ptr noundef nonnull align 1 dereferenceable(1) %1) [ "ptrauth"(i32 0, i64 9607) ]
  call void ptrauth (ptr @_ZN1C2fnEv, i32 0, i64 62113)(ptr noundef nonnull align 1 dereferenceable(1) %1) #3 [ "ptrauth"(i32 0, i64 62113) ]
```

Note that in the first case we authenticate function pointer with wrong discriminator, no resign is inserted.

This looks similar to https://cplusplus.github.io/CWG/issues/2486.html, though I do not know what is the status of this DR in clang.

**Comment by asl** [comment]
Tagging @rjmccall @ahmedbougacha

**Comment by asl** [comment]
Also tagging @ahatanak and @ojhunt

**Comment by Endilll** [comment]
> This looks similar to https://cplusplus.github.io/CWG/issues/2486.html, though I do not know what is the status of this DR in clang.

I opened #107131

**Comment by rjmccall** [comment]
I don't think the DR makes a difference to this bug — we have all the information we need statically in order to compile this correctly.

Is the member function pointer conversion reflected reasonably in the AST?  I could believe that either we don't have a proper `CastKind` for this or that the IRGen around it assumes that of course it's a no-op.

The ideal fix for this would do two things:
- make sure that we resign when we see this conversion in general
- emit member pointer constants specially when they're emitted in a converting context

We also presumably need to handle this properly in constant emission.

**Comment by asl** [comment]
> Is the member function pointer conversion reflected reasonably in the AST? I could believe that either we don't have a proper `CastKind` for this or that the IRGen around it assumes that of course it's a no-op.

There is implicit cast in AST as far as I can see:
```
ImplicitCastExpr <col:11, col:15> 'void (C::*)()' <NoOp>
`-UnaryOperator <col:11, col:15> 'void (C::*)() noexcept' prefix '&' cannot overflow
  `-DeclRefExpr <col:12, col:15> 'void () noexcept' CXXMethod 0xd57ffb0 'fn' 'void () noexcept'
    `-NestedNameSpecifier TypeSpec 'C'
```

So, probably IRGen is missing some cases?

FWIW, the issue exists in Apple clang as well:
```c++
@_ZN1C2fnEv.ptrauth = private unnamed_addr constant { ptr, i32, i64, i64 } { ptr @_ZN1C2fnEv, i32 0, i64 0, i64 62113 }, section "llvm.ptrauth", align 8

; Function Attrs: mustprogress ssp uwtable(sync)
define void @_Z3refv() local_unnamed_addr #0 {
  %1 = alloca %struct.C, align 1
  call void @llvm.lifetime.start.p0(i64 1, ptr nonnull %1) #3
  call void @_ZN1C2fnEv.ptrauth(ptr noundef nonnull %1) [ "ptrauth"(i32 0, i64 9607) ]
  call void @_ZN1C2fnEv(ptr noundef nonnull %1) #3
  call void @llvm.lifetime.end.p0(i64 1, ptr nonnull %1) #3
  ret void
}
```
the discriminators does not match in the operand bundle and wrapper global

**Comment by ahatanak** [comment]
It looks like `ScalarExprEmitter::VisitCastExpr` isn't resigning the pointer.

```
case CK_NoOp: {
    return CE->changesVolatileQualification() ? EmitLoadOfLValue(CE)
                                              : Visit(const_cast<Expr *>(E));
  }
```

`E` is the subexpression of the `ImplicitCastExpr`.

**Comment by asl** [comment]
Another interesting case:

```c++
#include <stdio.h>
#include <stdarg.h>

void NoexceptFunction() noexcept {
    printf("NE function\n");
}

[[gnu::noinline]]
void CallThroughThrowingFnPointer(int dummy, ...) {
    va_list args;
    va_start(args, dummy);
    auto* fn = va_arg(args, void(*)()); // non-noexcept function pointer
    fn();
    va_end(args);
}

int main() {
    CallThroughThrowingFnPointer(0, &NoexceptFunction); // no resigning code generated here (compiler can't know how the argument will be used)
}
```

**Comment by asl** [comment]
The second case was discussed at LLVM Developers Meeting. And it seems that the outcome is that `noexcept` should not be part of type discriminator as function pointers with and without `noexcept` are ABI compatible.

**Comment by rjmccall** [comment]
Right.  The standard has some rules here — e.g. making a call to a function well-defined if the signature differs only by `noexcept`-ness — that, even if incomplete, it's reasonable to read as trying to make the types have equivalent ABI.  That would imply that type discriminators should match exactly as well.

An additional pragmatic argument is that there is no way to declare a function pointer `noexcept` in C, and yet a function pointer exposed to C should never throw, and so it would be reasonable for C++ code working with that function pointer to declare it `noexcept`.  Note that function pointers don't actually have this problem, however, because they use a much looser discriminator algorithm that already ignores `noexcept`; only member function pointers use this strict rule based on the exact Itanium type mangling.
